src/org/sosy_lab/cpachecker/cfa/parser/llvm/CFABuilder.java

```diff
@@ -1215,8 +1226,12 @@ private CRightHandSide getConstant(final Value pItem, CType pExpectedType, final
         return funcId;
       }
 
-    } else if (pItem.isGlobalConstant() && pItem.isGlobalVariable()) {
-      return getAssignedIdExpression(pItem, pExpectedType, pFileName);
+    } else if (pItem.isGlobalVariable()) {
+        if (!pItem.isExternallyInitialized() || pItem.isGlobalConstant()) {
+            return getAssignedIdExpression(pItem, pExpectedType, pFileName);
+        } else {
+            throw new UnsupportedOperationException("LLVM parsing does not support this global variable: " + pItem);
```

Comment:
Maybe this is a naive question, but what other kinds of global variables exist? Can we tell the user more about what kind of global variables aren't working (in the exception message)?

Reply:
I don't think there are any other options. If the exception is thrown, then the global variable is uninitialized non-constant. I do not know about any other "type" of global variable.

Reply:
Hm, I'm a bit confused by the logic.

First, I'm confused about the branching above:
The branch conditions say that if a global variable is either locally initialized or a global constant,
a new id-expression is created; otherwise, an exception is thrown.
But don't we also want to create an id-expression if the global variable is externally initialized, but non-constant?
This case will, if I understand the method correctly, never happen (see below), but I'm still confused - there must be some good reason why you added this condition.

Second, the method is supposed to only handle constant values, so `pItem.isGlobalConstant()` must always be true at this point, right?
If that's the case, I would either keep it as a strong condition `pItem.isGlobalConstant() && ..` so that the UnsupportedOperationException below is provoked,
or put it into an assert/throw an AssertionError if it is not.

Reply:
I am not sure... the previous condition was `pItem.isGlobalConstant() && pItem.isGlobalVariable()`, so if `pItem.isGlobalConstant()` would be constantly true here, why would adding a disjunction with `!pItem.isExternallyInitialized()` change something?

Reply:
> but it had

Alright, that's good enough of an explanation/reason for me ;-)

> (I do not remember exactly what was the fix for, but I can try reverting the changes and find out).

That would be nice, because people in the future may assume that `pItem.isGlobalConstant()` is always true and add an regression by changing this.
But if it takes a lot of time to find out about the reason, we could also just add a warning comment like `// global constant may be false` (though I don't like that specific one).
Or is it obvious to you that `isGlobalConstant` may be wrong? Maybe I'm just missing something.

Reply:
Sorry for that looong pause in the conversation (and sorry for the long text that follows :). I finally got to track this down.  The problem that I was fixing is that if you have a global variable, it is statically initialized (if not initialized externally). Therefore, it has a value during compilation and you can use it as an initializer for some other global variable.
Here's the stack trace of the error that is solved by the patch:

```
Exception in thread "main" java.lang.UnsupportedOperationException: LLVM parsing does not support constant @calls_0 = common local_unnamed_addr global i32 0, align 4
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.getConstant(CFABuilder.java:1243)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.getConstant(CFABuilder.java:1176)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.getConstantAggregateInitializer(CFABuilder.java:1275)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.visitGlobalItem(CFABuilder.java:1875)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.iterateOverGlobals(CFABuilder.java:219)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.visit(CFABuilder.java:188)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.CFABuilder.build(CFABuilder.java:174)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.LlvmParser.buildCfa(LlvmParser.java:110)
	at org.sosy_lab.cpachecker.cfa.parser.llvm.LlvmParser.parseFile(LlvmParser.java:69)
	at org.sosy_lab.cpachecker.cfa.CFACreator.parseToCFAs(CFACreator.java:641)
	at org.sosy_lab.cpachecker.cfa.CFACreator.parseFileAndCreateCFA(CFACreator.java:433)
	at org.sosy_lab.cpachecker.core.CPAchecker.parse(CPAchecker.java:500)
	at org.sosy_lab.cpachecker.core.CPAchecker.run(CPAchecker.java:358)
	at org.sosy_lab.cpachecker.cmdline.CPAMain.main(CPAMain.java:176)
```

Since the `@calls_0` is not a global constant, the condition `pItem.isGlobalConstant() && pItem.isGlobalVariable()` is not satisfied and we get this exception. However, `pItem.isConstant()` is satisfied as global variables are constants (i.e. their address is constant, but the contents they point to may not be constant and that's what is checked by `isGlobalConstant()`). So the fix here asks whether the constant is global variable and if so, it check whether the value is known during compilation (i.e. whether the global is not externally initialized or is a global constant).

Reply:
Thanks for the explanation!